smarts/core/traffic_history.py

```diff
@@ -0,0 +1,139 @@
+        with nullcontext(self._db_cnxn) if self._db_cnxn else closing(
+            sqlite3.connect(self._db)
+        ) as dbcnxn:
+            cur = dbcnxn.cursor()
+            cur.execute(query, params)
+            row = cur.fetchone()
+            cur.close()
+        if not row:
+            return None
+        return row if result_type is tuple else result_type(row[0])
+
+    def _query_list(self, query: str, params=()):
+        with nullcontext(self._db_cnxn) if self._db_cnxn else closing(
+            sqlite3.connect(self._db)
+        ) as dbcnxn:
+            cur = dbcnxn.cursor()
+            for row in cur.execute(query, params):
+                yield row
+            cur.close()
+
+    @cached_property
+    def lane_width(self) -> float:
+        query = "SELECT value FROM Spec where key='map_net.lane_width'"
+        return self._query_val(float, query)
+
+    @cached_property
+    def target_speed(self) -> float:
+        query = "SELECT value FROM Spec where key='speed_limit_mps'"
+        return self._query_val(float, query)
+
+    @lru_cache(maxsize=32)
+    def vehicle_final_exit_time(self, vehicle_id: str) -> float:
+        query = "SELECT max(sim_time) FROM Trajectory WHERE vehicle_id = ?"
+        return self._query_val(float, query, params=(vehicle_id,))
+
+    def first_seen_times(self) -> Generator[Tuple[str, float], None, None]:
+        # For now, limit agent missions to just cars (V.type = 2)
+        query = """SELECT T.vehicle_id, min(T.sim_time)
+            FROM Trajectory AS T INNER JOIN Vehicle AS V ON T.vehicle_id=V.id
+            WHERE V.type = 2
+            GROUP BY vehicle_id"""
+        return self._query_list(query)
+
+    def vehicle_pose_at_time(
+        self, vehicle_id: str, sim_time: float
+    ) -> Tuple[float, float, float]:
+        query = """SELECT position_x, position_y, heading_rad
+                   FROM Trajectory
+                   WHERE vehicle_id = ? and sim_time = ?"""
+        return self._query_val(tuple, query, params=(int(vehicle_id), float(sim_time)))
+
+    def vehicle_ids_active_between(
+        self, start_time: float, end_time: float
+    ) -> Generator[int, None, None]:
+        query = "SELECT DISTINCT vehicle_id FROM Trajectory WHERE ? <= sim_time AND sim_time <= ?"
+        return self._query_list(query, (start_time, end_time))
+
+    def vehicles_active_between(
+        self, start_time: float, end_time: float
+    ) -> Generator[
+        Tuple[int, int, float, float, float, float, float, float], None, None
+    ]:
+        query = """SELECT V.id, V.type, V.length, V.width,
+                          T.position_x, T.position_y, T.heading_rad, T.speed
+                   FROM Vehicle AS V INNER JOIN Trajectory AS T ON V.id = T.vehicle_id
+                   WHERE T.sim_time > ? AND T.sim_time <= ?
+                   ORDER BY T.sim_time DESC"""
+        return self._query_list(query, (start_time, end_time))
```

Comment:
I would consider wrapping the generator with another generator to get types that way you do not need to wrap it later.
```suggestion
        row = self._query_list(query, (start_time, end_time))
        return (TrafficHistoryProvider._HistoryRow(*row) for row in rows)
```

Reply:
Good call.  Done.